### PR TITLE
UI test framework

### LINKED ISSUE #87: Test framework
Create a test framework for codegen, UI, error messages

### PR DESCRIPTION
**Pull request by ayazhafiz**
This commit adds a framework for testing the validity of hgb's
operations on GB programs by comparing UI outputs to checked-in golden
files. It's just a golden test basically. This commit demonstrates the
functionality with UI tests for TypeScript codegen, though we can easily
add more tests later on for error messages, other codegen, etc.

Closes #87

### CODE REVIEW

**Comment by ayazhafiz** [comment]
@lukasberglund  :)

File: src/Grammar/TypeScript.hs [hunk 1]
```diff
@@ -12,6 +13,10 @@ import Data.List (intercalate)
 import Grammar (Grammar, MapGrammar, (\>))
 import qualified Grammar as Hgb
 
+-- Generates TypeScript code from a GB program.
+codegen :: [Hgb.Statement] -> String
+codegen stmts = intercalate "" $ map (\line -> show line ++ "\n") (map (\>) stmts :: [Statement])
```
Comment (lukasberglund) [inline src/Grammar/TypeScript.hs:18]:
```suggestion
codegen = intercalate "\n" . map (show . (\>))
```
Could something like this work?

Comment (lukasberglund) [reply src/Grammar/TypeScript.hs:18]: Or `codegen = intercalate "\n" . map show . map (\>)`

File: test/ui/UITest.hs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import Test.Tasty.Golden.Advanced
+import Test.Tasty.HUnit
+
+import qualified Grammar as GB (Statement)
+import Grammar (Grammar, (\>))
+import qualified Grammar.TypeScript as TS (codegen)
+import Lexer
+import Parser
+
+main :: IO ()
```
Comment (lukasberglund) [inline test/ui/UITest.hs:27]: This is far out dude. Nice.

**Review by lukasberglund — APPROVED** [review]
Looks great. Looks like it wasn't easy. I just have one suggestion. You can take it or leave it :)

File: src/Grammar/TypeScript.hs (hunk 1 above)
Comment (ayazhafiz) [reply src/Grammar/TypeScript.hs]: done

**Pull request merged**